openerp/addons/base/ir/ir_qweb/qweb.py

```diff
@@ -0,0 +1,1372 @@
+    # FIXME: maybe values could provide a ChainMap-style clone?
+    for k in base_ctx.keys():
+        base_ctx[k] = ctx[k]
+
+_FORMAT_REGEX = re.compile(
+    '(?:'
+        # ruby-style pattern
+        '#\{(.+?)\}'
+    ')|(?:'
+        # jinja-style pattern
+        '\{\{(.+?)\}\}'
+    ')')
+
+
+class QWebField(object):
+    def attributes(self, record, field_name, options, values=None):
+        return {}
+    def record_to_html(self, record, field_name, options, values=None):
+        return getattr(record, field_name, None)
+
+
+class frozendict(dict):
+    """ An implementation of an immutable dictionary. """
+    def __delitem__(self, key):
+        raise NotImplementedError("'__delitem__' not supported on frozendict")
+    def __setitem__(self, key, val):
+        raise NotImplementedError("'__setitem__' not supported on frozendict")
+    def clear(self):
+        raise NotImplementedError("'clear' not supported on frozendict")
+    def pop(self, key, default=None):
+        raise NotImplementedError("'pop' not supported on frozendict")
+    def popitem(self):
+        raise NotImplementedError("'popitem' not supported on frozendict")
+    def setdefault(self, key, default=None):
+        raise NotImplementedError("'setdefault' not supported on frozendict")
+    def update(self, *args, **kwargs):
+        raise NotImplementedError("'update' not supported on frozendict")
+    def __hash__(self):
+        return hash(frozenset((key, freehash(val)) for key, val in self.iteritems()))
+
+
+####################################
+###             QWeb             ###
+####################################
+
+
+class QWeb(object):
+
+    _void_elements = frozenset([
+        'area', 'base', 'br', 'col', 'embed', 'hr', 'img', 'input', 'keygen',
+        'link', 'menuitem', 'meta', 'param', 'source', 'track', 'wbr'])
+    _name_gen = count()
+
+    def render(self, template, values=None, **options):
+        """
+        'options' can be used by QWeb methods
+        'values' is used to evaluate template values
+        """
+        body = []
+        self.compile(template, options)(self, body.append, values)
+        return u''.join(body)
+
+    def compile(self, template, options):
+        """
+        Return a "qweb function"
+        * parameters: ``append`` ``values``
+        * use append method to add content
+        """
+        if options is None:
+            options = {}
+
+        _options = dict(options)
+        options = frozendict(options)
+
+        element, document = self.get_template(template, options)
+        name = element.get('t-name', 'unknown')
+
+        _options['template'] = template
+        _options['ast_calls'] = []
+        _options['root'] = element.getroottree()
+        _options['last_path_node'] = None
+
+        # generate ast
+
+        astmod = self._base_module()
+        try:
+            self._call_body(_options, ['empty'], prefix='template_%s' % name.replace('.', '_'))
```

Comment:
Using `_call_body` is a bit confusing here: we do not need the ast of the call, we only need the function definition.
I would make a method to generate the ast of the function definition, and let `_call_body` use it. That would avoid the body patching below.